pkg/activator/throttler.go

```diff
@@ -0,0 +1,212 @@
+type Throttler struct {
+	breakers      map[RevisionID]*queue.Breaker
+	breakerParams queue.BreakerParams
+	logger        *zap.SugaredLogger
+	getEndpoints  func(RevisionID) (int32, error)
+	getRevision   func(RevisionID) (*v1alpha1.Revision, bool, error)
+	mux           sync.Mutex
+}
+
+// Remove deletes the breaker from the bookkeeping.
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity updates the max concurrency of the Breaker corresponding to a revision.
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	revision, err := t.revision(rev)
+	if err != nil {
+		return err
+	}
+	breaker, _ := t.getOrCreateBreaker(rev)
+	err = t.updateCapacity(revision, breaker, size)
+	return err
+}
+
+// Try potentially registers a new breaker in our bookkeeping
+// and executes the `function` on the Breaker.
+// It returns an error if either breaker doesn't have enough capacity,
+// or breaker's registration didn't succeed, e.g. getting endpoints or update capacity failed.
+func (t *Throttler) Try(rev RevisionID, function func()) error {
+	t.mux.Lock()
+	breaker, existed := t.getOrCreateBreaker(rev)
+	if !existed {
+		err := t.forceUpdateCapacity(rev, breaker)
+		if err != nil {
+			return err
+		}
+	}
+	t.mux.Unlock()
+	if ok := breaker.Maybe(function); !ok {
+		return errors.New(OverloadMessage)
+	}
+	return nil
+}
+
+// This method updates Breaker's concurrency and requires external synchronization, `updateCapacity` requires `mux` to be held.
+func (t *Throttler) updateCapacity(revision *v1alpha1.Revision, breaker *queue.Breaker, size int32) (err error) {
+	cc := int32(revision.Spec.ContainerConcurrency)
+	// The concurrency is unlimited, thus hand out as many tokens as we can in this breaker.
+	if cc == 0 {
+		cc = t.breakerParams.MaxConcurrency
+	}
+	delta := size*cc - breaker.Capacity()
+	// Do not update throttler's concurrency if the same number of endpoints is received
+	// or the number is negative.
+	if delta <= 0 {
+		return nil
+	}
+	if err := breaker.UpdateConcurrency(delta); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getOrCreateBreaker retrieves existing breaker or creates a new one.
+// This is important for not loosing the update signals
+// that came before the requests reached the Activator's Handler.
+func (t *Throttler) getOrCreateBreaker(rev RevisionID) (*queue.Breaker, bool) {
+	breaker, ok := t.breakers[rev]
+	if !ok {
+		breaker = queue.NewBreaker(t.breakerParams)
+		t.breakers[rev] = breaker
+	}
+	return breaker, ok
+}
+
+// revision returns the Revision id if it exists.
+// Do not inline it inside `updateCapacity` to save the deletion
+// of newly created breaker if no revision exists yet.
+// TODO: this method must be inlined in t.getRevision() when the throttler is enabled
```

Comment:
Why don't we assume it's inlined already? Does the rest of our current code care at all?

Reply:
Agree - updated.